### DISCUSSION TITLE
can I disable encryption while testing?

### DISCUSSION THREAD

**Discussion by Winters123** [microsoft/msquic#2671] [Q&A] [answered]
Hi, I'm wondering if I can disable encryption while testing. If so, what should I do? 

BTW, I noticed this in the doc: `Important - ChaCha20-Poly1305 is not yet supported with MsQuic and Schannel, so this doesn't do anything yet.` Does it mean that I could bypass encryption when I choose to use ChaCha20-Poly1305?

**Comment by nibanks** [comment, marked as answer]
No, you can't leverage unsupported ChaCha20-Poly1305 to just disable encryption. It doesn't work at all, so it won't even let you try to use it. We have a flag that leverages a QUIC extension I wrote to disable QUIC encryption **after** the handshake, but I suspect it's not going to help you. It might be best if you start a new question asking to the core of what you're trying to accomplish and we can follow up there.

**Reply by wtao0221** [reply]
Hi Nick,
Which QUIC extension are you referring to? Do you have any doc on this?
Thanks!

**Reply by nibanks** [reply]
It is linked to on our main readme, but here it is too: https://tools.ietf.org/html/draft-banks-quic-disable-encryption. As you can see, I wrote it, so I can answer any questions you might have on it.

**Reply by wtao0221** [reply]
Thanks for the link! Yeah I saw it. One quick question, does it mean during the handshake between client and server, in the TLS extension, the client indicates that the following application data will not be encrypted, and server will accept it, which is supported in msquic. Is that correct?

**Reply by nibanks** [reply]
Yes, both sides have to be configured and agree to use the extension to negotiate it during the handshake. Then after the handshake, encryption is disabled.

**Reply by wtao0221** [reply]
Oh, I see it now https://github.com/microsoft/msquic/blob/main/docs/Settings.md

**Reply by nibanks** [reply]
Take a look at `QUIC_PARAM_CONN_DISABLE_1RTT_ENCRYPTION` [here](https://github.com/microsoft/msquic/blob/main/src/inc/msquic.h#L752).

**Reply by wtao0221** [reply]
Yeah, so basically, i insert SetParam() when client initializes a connection (https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c#L813) to the server, and when server accepts a connection (https://github.com/microsoft/msquic/blob/main/src/tools/sample/sample.c#L378).

It seems that the data is still encrypted. Maybe i am wrong with the APIs?

**Reply by nibanks** [reply]
How are you determining it's still encrypted? Wireshark doesn't support this extension, so it doesn't show it as decrypted.

**Reply by wtao0221** [reply]
I sent the "hello world" message and used tcpdump to check the packet.

If the payload is not encrypted, I may be able to see the plain text?

BTW, do you have any recommendation of tools inspecting the packets?

**Reply by nibanks** [reply]
I recommend taking a look at https://github.com/microsoft/msquic/blob/main/docs/Diagnostics.md and grab the logs to see what's going on.

**Reply by wtao0221** [reply]
Thanks!

Interestingly, i think i overlapped with some sizeof(QUIC_BUFFER) bytes. (what is it for?)

BTW, is there any example showing how to read received bytes from a stream? I saw the struct of STREAM_EVENT only has number of bytes received, but i am not sure where the buffer is.

**Reply by nibanks** [reply]
https://github.com/microsoft/msquic/blob/main/docs/Streams.md should have the details you need. If it's missing something, please open up an issue with what you'd like to see in it, with as much detail as possible.
